File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/command/imagesearch/provider/google/GoogleCustomSearchProvider.java
package com.github.bumblebee.command.imagesearch.provider.google;

import com.github.bumblebee.command.imagesearch.domain.Image;
import com.github.bumblebee.command.imagesearch.domain.ImageProvider;
import com.github.bumblebee.command.imagesearch.provider.google.response.GoogleCustomSearchResponse;
import feign.Feign;
import feign.gson.GsonDecoder;
import feign.slf4j.Slf4jLogger;
import org.apache.commons.io.FilenameUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.util.CollectionUtils;

import java.util.List;
import java.util.stream.Collectors;

@Component
public class GoogleCustomSearchProvider implements ImageProvider {

    private static final Logger log = LoggerFactory.getLogger(GoogleCustomSearchProvider.class);

    private final GoogleSearchApi searchApi;
    private final GoogleCustomSearchConfig config;

    @Autowired
    public GoogleCustomSearchProvider(GoogleCustomSearchConfig config) {
        this.config = config;
        this.searchApi = Feign.builder()
                .decoder(new GsonDecoder())
                .logLevel(feign.Logger.Level.BASIC)
                .logger(new Slf4jLogger())
                .target(GoogleSearchApi.class, GoogleSearchApi.API_ROOT);
    }

    @Override
    public List<Image> search(String query) {
        GoogleCustomSearchResponse response = searchApi.queryPictures(query, config.getKey(), config.getCustomSearchId());
        if (response != null && !CollectionUtils.isEmpty(response.getItems())) {
            log.info("Google found {} results for: {}", response.getItems().size(), query);
            return response.getItems()
                    .stream()
                    .filter(item -> !FilenameUtils.getExtension(item.getLink()).isEmpty())
                    .collect(Collectors.toList());
        }
        return null;
    }

}

File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/command/imagesearch/provider/google/GoogleCustomSearchProvider.kt
package com.github.bumblebee.command.imagesearch.provider.google

import com.github.bumblebee.command.imagesearch.domain.Image
import com.github.bumblebee.command.imagesearch.domain.ImageProvider
import com.github.bumblebee.util.logger
import feign.Feign
import feign.gson.GsonDecoder
import feign.slf4j.Slf4jLogger
import org.apache.commons.io.FilenameUtils
import org.springframework.stereotype.Component

@Component
class GoogleCustomSearchProvider(private val config: GoogleCustomSearchConfig) : ImageProvider {

    private val searchApi: GoogleSearchApi = Feign.builder()
            .decoder(GsonDecoder())
            .logLevel(feign.Logger.Level.BASIC)
            .logger(Slf4jLogger())
            .target(GoogleSearchApi::class.java, GoogleSearchApi.API_ROOT)

    override fun search(query: String): List<Image> {
        val response = searchApi.queryPictures(query, config.key, config.customSearchId)
        val items = response.items ?: emptyList()
        return if (items.isNotEmpty()) {
            log.info("Google found {} results for: {}", items.size, query)

            return items.filter { item -> !FilenameUtils.getExtension(item.link).isEmpty() }
        } else emptyList()
    }

    companion object {
        private val log = logger<GoogleCustomSearchProvider>()
    }
}
